// stratum-pool/lib/apiClient.js
"use strict";

const request = require("request");
const rp = require("request-promise");

this.ip = "8.9.31.244";
var events = require("events");

process.env["NODE_TLS_REJECT_UNAUTHORIZED"] = 0;

/* todo 
add logger support(events)
low timeout
sentry
logger
*/
class APIClient {
  constructor(ip) {
    this.ip = ip;
    //this.logger = function(severity, message) {
    //    console.log(severity + ": " + message);
    //};

    // Check if online?
  }

  // TODO SET LOW TIMEOUT!!!
  // TODO handle errors
  cmd(url, callback) {
    //console.log("cmd url:" + url);
    request(url, { json: true }, (err, res, body) => {
      if (err) {
        callback(error, null);
        //return console.log(err);
      } else {
        //console.log("res.body:" + JSON.stringify(res.body));
        callback(null, res.body);
      }
    });
  }

  getCut2() {
    return rp(`https://${this.ip}/chainweb/0.0/mainnet01/cut`).then(body => {
      let response = JSON.parse(body);
      return body;
    });
  }
  getCut() {
    let req = request(
      `https://${this.ip}/chainweb/0.0/mainnet01/cut`,
      (err, res, body) => {
        if (err)
          console.log("Errror in getHeight: " + err);
        //console.log(`body:JSON.stringify(${body})`);
        return body;
      });
  };

  getPeerCount() {
    console.log("getPeerCount");
    const res = this.cmd(
      `https://${this.ip}/chainweb/0.0/mainnet01/cut/peer`,
      function(err, data) {
        if (err) console.log("Errror in getHeight: " + err);
        //console.log("getPeerCount() " + Object.keys(data).length);
        return data.length;
      }
    );
  }

  getDifficulty() {
    console.log("returning 12345");
    return 12345;
  }

  getHeaders() {
    console.log("getHeaders");
    const res = this.cmd(
      `https://${this.ip}/chainweb/0.0/mainnet01/headers`,
      function(err, data) {
        if (err) console.log("Errror in getHeaders: " + err);
        return;
        data.height;
      }
    );
  }

  getHeight() {
    console.log("getHeight");
    const res = this.cmd(
      `https://${this.ip}/chainweb/0.0/mainnet01/cut`,
      function(err, data) {
        if (err)
          console.log("Errror in getHeight: " + err);
        return data.height;
      }
    );
  }
}

module.exports = APIClient;
